File: src/main/java/com/da/itdaing/domain/geo/service/GeoZoneService.java
// src/main/java/com/da/itdaing/domain/geo/service/GeoZoneService.java
package com.da.itdaing.domain.geo.service;

import com.da.itdaing.domain.common.enums.AreaStatus;
import com.da.itdaing.domain.common.enums.ZoneStatus;
import com.da.itdaing.domain.geo.dto.GeoDtos.*;
import com.da.itdaing.domain.geo.entity.ZoneArea;
import com.da.itdaing.domain.geo.entity.ZoneCell;
import com.da.itdaing.domain.geo.repository.ZoneAreaRepository;
import com.da.itdaing.domain.geo.repository.ZoneCellRepository;
import com.da.itdaing.domain.user.entity.Users;
import com.da.itdaing.domain.user.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.locationtech.jts.geom.Coordinate;
import org.locationtech.jts.geom.Envelope;
import org.locationtech.jts.geom.Geometry;
import org.locationtech.jts.geom.GeometryFactory;
import org.locationtech.jts.geom.Point;
import org.locationtech.jts.geom.prep.PreparedGeometry;
import org.locationtech.jts.geom.prep.PreparedGeometryFactory;
import org.springframework.data.domain.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.locationtech.jts.geom.*;

@Service
@RequiredArgsConstructor
@Transactional
public class GeoZoneService {

    private final ZoneCellRepository zoneRepo;
    private final ZoneAreaRepository areaRepo;
    private final UserRepository userRepo;

    /** 판매자: 존 생성 (기본 PENDING) + 폴리곤 포함검사(강제) */
    public ZoneResponse createZone(Long sellerId, CreateZoneRequest req) {
        Users owner = userRepo.findById(sellerId).orElseThrow();
        ZoneArea area = areaRepo.findById(req.getAreaId()).orElseThrow();

        // 구역 상태 체크
        if (area.getStatus() == AreaStatus.HIDDEN || area.getStatus() == AreaStatus.UNAVAILABLE) {
            throw new IllegalStateException("해당 구역은 현재 존 등록이 불가합니다.");
        }

        // 포함검사: (lng, lat) 좌표가 area.polygonGeoJson 내부(또는 경계선)에 존재해야 함
        validatePointInsideOrThrow(area.getPolygonGeoJson(), req.getLng(), req.getLat());

        ZoneCell z = ZoneCell.builder()
            .zoneArea(area)
            .owner(owner)
            .label(req.getLabel())
            .detailedAddress(req.getDetailedAddress())
            .lat(req.getLat())
            .lng(req.getLng())
            .status(ZoneStatus.PENDING)
            .maxCapacity(req.getMaxCapacity())
            .notice(req.getNotice())
            .build();

        zoneRepo.save(z);
        return toDto(z);
    }

    /** 관리자: 특정 구역의 존 목록 조회 */
    @Transactional(readOnly = true)
    public ZoneListResponse listZonesByArea(Long areaId, int page, int size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "id"));
        Page<ZoneCell> data = zoneRepo.findByZoneArea_Id(areaId, pageable);
        return pageToList(data);
    }

    /** 판매자: 내가 만든 존 목록 조회 */
    @Transactional(readOnly = true)
    public ZoneListResponse listMyZones(Long sellerId, int page, int size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "id"));
        Page<ZoneCell> data = zoneRepo.findByOwner_Id(sellerId, pageable);
        return pageToList(data);
    }

    /** 관리자: 존 상태 변경(승인/반려/숨김 등) */
    public void changeZoneStatus(Long zoneId, ZoneStatus status) {
        ZoneCell z = zoneRepo.findById(zoneId).orElseThrow();
        z.changeStatus(status);
        zoneRepo.save(z);
    }

    /* ---------- 포함검사 유틸 (GeoJSON Polygon/MultiPolygon, 좌표는 WGS84 [lng, lat]) ---------- */

    private void validatePointInsideOrThrow(String polygonGeoJson, double lng, double lat) {
        if (polygonGeoJson == null || polygonGeoJson.isBlank()) {
            throw new IllegalArgumentException("구역 폴리곤(GeoJSON)이 없습니다.");
        }
        try {
            Geometry geom = readGeometryFromGeoJson(polygonGeoJson); // ← GeoJSON → JTS

            Envelope env = geom.getEnvelopeInternal();
            if (!env.contains(new Coordinate(lng, lat))) {
                throw new IllegalArgumentException("선택 좌표는 구역 외부(BBOX)입니다.");
            }

            GeometryFactory gf = new GeometryFactory();
            Point p = gf.createPoint(new Coordinate(lng, lat));
            PreparedGeometry prepared = PreparedGeometryFactory.prepare(geom);

            if (!prepared.covers(p)) { // 경계 포함 허용
                throw new IllegalArgumentException("선택 좌표는 구역 폴리곤 내부가 아닙니다.");
            }
        } catch (IllegalArgumentException e) {
            throw e;
        } catch (Exception e) {
            throw new IllegalArgumentException("폴리곤 파싱/검증 실패(GeoJSON/WGS84 [lng,lat] 확인 필요).", e);
        }
    }

    /** 최소 지원: Polygon / MultiPolygon (Feature/FeatureCollection은 첫 geometry만 처리) */
    private Geometry readGeometryFromGeoJson(String geojson) throws Exception {
        ObjectMapper om = new ObjectMapper();
        JsonNode root = om.readTree(geojson);

        // Feature / FeatureCollection 처리(간단히 첫 geometry만)
        if (root.has("type") && "Feature".equalsIgnoreCase(root.get("type").asText()) && root.has("geometry")) {
            root = root.get("geometry");
        } else if (root.has("type") && "FeatureCollection".equalsIgnoreCase(root.get("type").asText())
            && root.has("features") && root.get("features").isArray() && root.get("features").size() > 0) {
            JsonNode first = root.get("features").get(0);
            root = first.has("geometry") ? first.get("geometry") : root;
        }

        String type = root.has("type") ? root.get("type").asText() : null;
        if (type == null) throw new IllegalArgumentException("GeoJSON type이 없습니다.");

        GeometryFactory gf = new GeometryFactory();

        if ("Polygon".equalsIgnoreCase(type)) {
            return polygonFromCoords(gf, root.get("coordinates"));
        } else if ("MultiPolygon".equalsIgnoreCase(type)) {
            return multiPolygonFromCoords(gf, root.get("coordinates"));
        } else {
            throw new IllegalArgumentException("지원하지 않는 GeoJSON 타입: " + type);
        }
    }

    private Polygon polygonFromCoords(GeometryFactory gf, JsonNode coordsNode) {
        if (coordsNode == null || !coordsNode.isArray() || coordsNode.isEmpty()) {
            throw new IllegalArgumentException("Polygon coordinates가 비어있습니다.");
        }
        // 외곽 링
        LinearRing shell = ringFromLinearCoords(gf, coordsNode.get(0));
        // 홀(내부 링들)
        LinearRing[] holes = new LinearRing[Math.max(0, coordsNode.size() - 1)];
        for (int i = 1; i < coordsNode.size(); i++) {
            holes[i - 1] = ringFromLinearCoords(gf, coordsNode.get(i));
        }
        return gf.createPolygon(shell, holes);
    }

    private MultiPolygon multiPolygonFromCoords(GeometryFactory gf, JsonNode coordsNode) {
        if (coordsNode == null || !coordsNode.isArray() || coordsNode.isEmpty()) {
            throw new IllegalArgumentException("MultiPolygon coordinates가 비어있습니다.");
        }
        Polygon[] polys = new Polygon[coordsNode.size()];
        for (int i = 0; i < coordsNode.size(); i++) {
            polys[i] = polygonFromCoords(gf, coordsNode.get(i));
        }
        return gf.createMultiPolygon(polys);
    }

    private LinearRing ringFromLinearCoords(GeometryFactory gf, JsonNode ringNode) {
        if (ringNode == null || !ringNode.isArray() || ringNode.size() < 4) {
            throw new IllegalArgumentException("LinearRing 좌표는 최소 4개([lng,lat]...)이어야 합니다.");
        }
        Coordinate[] coords = new Coordinate[ringNode.size()];
        for (int i = 0; i < ringNode.size(); i++) {
            JsonNode p = ringNode.get(i);
            double lng = p.get(0).asDouble();
            double lat = p.get(1).asDouble();
            coords[i] = new Coordinate(lng, lat);
        }
        // 폐합 보정: 첫 점 != 마지막 점이면 닫아준다
        if (!coords[0].equals2D(coords[coords.length - 1])) {
            Coordinate[] closed = new Coordinate[coords.length + 1];
            System.arraycopy(coords, 0, closed, 0, coords.length);
            closed[closed.length - 1] = coords[0];
            coords = closed;
        }
        return gf.createLinearRing(coords);
    }

    /* ---------- helpers ---------- */

    private ZoneResponse toDto(ZoneCell z) {
        return ZoneResponse.builder()
            .id(z.getId())
            .areaId(z.getZoneArea().getId())
            .ownerId(z.getOwner().getId())
            .label(z.getLabel())
            .detailedAddress(z.getDetailedAddress())
            .lat(z.getLat())
            .lng(z.getLng())
            .status(z.getStatus())
            .maxCapacity(z.getMaxCapacity())
            .notice(z.getNotice())
            .createdAt(z.getCreatedAt())
            .updatedAt(z.getUpdatedAt())
            .build();
    }

    private ZoneListResponse pageToList(Page<ZoneCell> page) {
        return ZoneListResponse.builder()
            .items(page.getContent().stream().map(this::toDto).toList())
            .totalElements(page.getTotalElements())
            .totalPages(page.getTotalPages())
            .page(page.getNumber())
            .size(page.getSize())
            .build();
    }
}

File: itdaing-app/src/services/geoZoneService.js
/**
 * Geo Zone/Area/Cell 관련 API 서비스
 * - 관리자: 존(zone_area) 생성/수정/삭제, 셀(zone_cell) 관리
 * - 판매자: 존 목록 조회, 셀 선택
 */

import { client } from '@/api/client';

const BASE_URL = '/api/geo';

/* ============ Area (존 영역) API ============ */

/**
 * 구역(Area) 목록 조회
 * @param {Object} params - { keyword, page, size }
 * @returns {Promise<{items: Array, totalElements: number, totalPages: number, page: number, size: number}>}
 */
export const listAreas = async ({ keyword = '', page = 0, size = 20 } = {}) => {
  const response = await client.get(`${BASE_URL}/areas`, {
    params: { keyword, page, size },
  });
  return response.data;
};

/**
 * 구역(Area) 상세 조회
 * @param {number} id - 구역 ID
 * @returns {Promise<Object>}
 */
export const getArea = async (id) => {
  const response = await client.get(`${BASE_URL}/areas/${id}`);
  return response.data;
};

/**
 * 구역(Area) 생성 (관리자 전용)
 * @param {Object} data - { name, polygonGeoJson, status, maxCapacity, notice, regionId }
 * @returns {Promise<Object>}
 */
export const createArea = async (data) => {
  const response = await client.post(`${BASE_URL}/areas`, data);
  return response.data;
};

/**
 * 구역(Area) 수정 (관리자 전용)
 * @param {number} id - 구역 ID
 * @param {Object} data - { name, polygonGeoJson, status, maxCapacity, notice, regionId }
 * @returns {Promise<Object>}
 */
export const updateArea = async (id, data) => {
  const response = await client.put(`${BASE_URL}/areas/${id}`, data);
  return response.data;
};

/**
 * 구역(Area) 삭제 (관리자 전용)
 * @param {number} id - 구역 ID
 * @returns {Promise<void>}
 */
export const deleteArea = async (id) => {
  await client.delete(`${BASE_URL}/areas/${id}`);
};

/* ============ Zone (존/셀) API ============ */

/**
 * 존(Zone) 생성 (관리자/판매자)
 * @param {Object} data - { areaId, ownerId, label, detailedAddress, geometryData, maxCapacity, notice }
 * @returns {Promise<Object>}
 */
export const createZone = async (data) => {
  const response = await client.post(`${BASE_URL}/zones`, data);
  return response.data;
};

/**
 * 내가 만든 존 목록 (판매자)
 * @param {Object} params - { page, size }
 * @returns {Promise<{items: Array, totalElements: number, totalPages: number, page: number, size: number}>}
 */
export const listMyZones = async ({ page = 0, size = 20 } = {}) => {
  const response = await client.get(`${BASE_URL}/zones/me`, {
    params: { page, size },
  });
  return response.data;
};

/**
 * 특정 구역의 존 목록 (관리자)
 * @param {number} areaId - 구역 ID
 * @param {Object} params - { page, size }
 * @returns {Promise<{items: Array, totalElements: number, totalPages: number, page: number, size: number}>}
 */
export const listZonesByArea = async (areaId, { page = 0, size = 20 } = {}) => {
  const response = await client.get(`${BASE_URL}/zones`, {
    params: { areaId, page, size },
  });
  return response.data;
};

/**
 * 존 상태 변경 (관리자)
 * @param {number} zoneId - 존 ID
 * @param {string} status - 상태 (PENDING, APPROVED, REJECTED, etc.)
 * @returns {Promise<void>}
 */
export const changeZoneStatus = async (zoneId, status) => {
  await client.patch(`${BASE_URL}/zones/${zoneId}/status`, { status });
};

/* ============ Cell (셀/부스) API ============ */

/**
 * 셀(Cell) 목록 조회
 * @param {Object} params - { areaId, page, size }
 * @returns {Promise<{items: Array, totalElements: number, totalPages: number, page: number, size: number}>}
 */
export const listCells = async ({ areaId, page = 0, size = 50 } = {}) => {
  const response = await client.get(`${BASE_URL}/cells`, {
    params: { areaId, page, size },
  });
  return response.data;
};

/**
 * 셀(Cell) 생성 (관리자 전용)
 * @param {Object} data - { areaId, ownerId, label, detailedAddress, geometryData, status, maxCapacity, notice }
 * @returns {Promise<Object>}
 */
export const createCell = async (data) => {
  const response = await client.post(`${BASE_URL}/cells`, data);
  return response.data;
};

/**
 * 셀(Cell) 수정 (관리자 전용)
 * @param {number} id - 셀 ID
 * @param {Object} data - { areaId, ownerId, label, detailedAddress, geometryData, status, maxCapacity, notice }
 * @returns {Promise<Object>}
 */
export const updateCell = async (id, data) => {
  const response = await client.put(`${BASE_URL}/cells/${id}`, data);
  return response.data;
};

/**
 * 셀(Cell) 삭제 (관리자 전용)
 * @param {number} id - 셀 ID
 * @returns {Promise<void>}
 */
export const deleteCell = async (id) => {
  await client.delete(`${BASE_URL}/cells/${id}`);
};

/* ============ 헬퍼 함수 ============ */

/**
 * GeoJSON Polygon 문자열을 파싱하여 좌표 배열로 변환
 * @param {string} geoJsonStr - GeoJSON Polygon 문자열
 * @returns {Array<{lat: number, lng: number}>} 좌표 배열
 */
export const parseGeoJsonPolygon = (geoJsonStr) => {
  if (!geoJsonStr) return [];
  
  try {
    const geo = JSON.parse(geoJsonStr);
    if (geo.type !== 'Polygon' || !geo.coordinates || !geo.coordinates[0]) {
      return [];
    }
    
    // GeoJSON은 [lng, lat] 순서이므로 변환
    return geo.coordinates[0].map(([lng, lat]) => ({ lat, lng }));
  } catch {
    return [];
  }
};

/**
 * 좌표 배열을 GeoJSON Polygon 문자열로 변환
 * @param {Array<{lat: number, lng: number}>} coords - 좌표 배열
 * @returns {string} GeoJSON Polygon 문자열
 */
export const toGeoJsonPolygon = (coords) => {
  if (!coords || coords.length === 0) return '';
  
  // GeoJSON은 [lng, lat] 순서
  const ring = coords.map(({ lat, lng }) => [lng, lat]);
  
  // 첫/마지막 좌표가 다르면 닫기
  const first = ring[0];
  const last = ring[ring.length - 1];
  if (first[0] !== last[0] || first[1] !== last[1]) {
    ring.push([...first]);
  }
  
  return JSON.stringify({
    type: 'Polygon',
    coordinates: [ring],
  });
};

export default {
  listAreas,
  getArea,
  createArea,
  updateArea,
  deleteArea,
  createZone,
  listMyZones,
  listZonesByArea,
  changeZoneStatus,
  listCells,
  createCell,
  updateCell,
  deleteCell,
  parseGeoJsonPolygon,
  toGeoJsonPolygon,
};

